graph/src/ipfs/gateway_client.rs

```diff
@@ -0,0 +1,702 @@
+use std::time::Duration;
+
+use anyhow::anyhow;
+use async_trait::async_trait;
+use bytes::Bytes;
+use bytes::BytesMut;
+use derivative::Derivative;
+use futures03::stream::BoxStream;
+use futures03::StreamExt;
+use futures03::TryStreamExt;
+use http::header::ACCEPT;
+use http::header::CACHE_CONTROL;
+use reqwest::StatusCode;
+use slog::Logger;
+
+use crate::derive::CheapClone;
+use crate::ipfs::retry_policy::retry_policy;
+use crate::ipfs::CanProvide;
+use crate::ipfs::Cat;
+use crate::ipfs::CatStream;
+use crate::ipfs::ContentPath;
+use crate::ipfs::GetBlock;
+use crate::ipfs::IpfsClient;
+use crate::ipfs::IpfsError;
+use crate::ipfs::IpfsResult;
+use crate::ipfs::ServerAddress;
+
+/// The request that verifies that the IPFS gateway is accessible is generally fast because
+/// it does not involve querying the distributed network.
+const TEST_REQUEST_TIMEOUT: Duration = Duration::from_secs(300);
+
+#[derive(Clone, CheapClone, Derivative)]
+#[derivative(Debug)]
+/// A client that connects to an IPFS gateway.
+///
+/// Reference: <https://specs.ipfs.tech/http-gateways/path-gateway>
+pub struct IpfsGatewayClient {
+    server_address: ServerAddress,
+
+    #[derivative(Debug = "ignore")]
+    http_client: reqwest::Client,
+
+    logger: Logger,
+    test_request_timeout: Duration,
+}
+
+impl IpfsGatewayClient {
+    /// Creates a new [IpfsGatewayClient] with the specified server address.
+    /// Verifies that the server is responding to IPFS gateway requests.
+    pub async fn new(server_address: impl AsRef<str>, logger: &Logger) -> IpfsResult<Self> {
+        let client = Self::new_unchecked(server_address, logger)?;
+
+        client
+            .send_test_request()
+            .await
+            .map_err(|reason| IpfsError::InvalidServer {
+                server_address: client.server_address.clone(),
+                reason,
+            })?;
+
+        Ok(client)
+    }
+
+    /// Creates a new [IpfsGatewayClient] with the specified server address.
+    /// Does not verify that the server is responding to IPFS gateway requests.
+    pub fn new_unchecked(server_address: impl AsRef<str>, logger: &Logger) -> IpfsResult<Self> {
+        Ok(Self {
+            server_address: ServerAddress::new(server_address)?,
+            http_client: reqwest::Client::new(),
+            logger: logger.to_owned(),
+            test_request_timeout: TEST_REQUEST_TIMEOUT,
+        })
+    }
+
+    pub fn into_boxed(self) -> Box<dyn IpfsClient> {
+        Box::new(self)
+    }
+
+    async fn send_test_request(&self) -> anyhow::Result<()> {
```

Comment:
Not sure what this is trying to achieve, is this only used for tests or is this a check for something specific? If it's only used for tests, could it be marked with cfg(test) ?

Reply:
Test requests are used to check which type of API a server address supports (RPC or Gateway) in order to initialize the appropriate client.